### PR TITLE
switch to pre-commit

### PR DESCRIPTION
**Pull request by akaszynski**
##### tl;dr
Let's use `pre-commit` to simplify our lives and make it easier and more fun to contribute.

### Problem

We're quite strict about our code standards, and that's great! We're writing consistent, readable code that many people feel comfortable contributing to. The problem with our existing approach is it's:

- not cross-platform (windows doesn't natively support makefiles)
- not automatic
- scattered across several makefile sections
- not intuitive

This makes it really, really annoying to ensure commits meet coding standards as one has to fire off one or more make commands and remember to do it.

### Proposed Solution

This PR suggests we use [pre-commit](https://pre-commit.com/) as done in [pandas/.pre-commit-config.yaml](https://github.com/pandas-dev/pandas/blob/main/.pre-commit-config.yaml) to:

- provide a cross-platform solution
- automate it with a simple to install pre-commit hook
- run automatically (and quickly) after each and every commit

`pre-commit` works on all major OSes, and is installed with:

```
pip install pre-commit
```

and then can be (manually) used with:

```
pre-commit run --all-files
```

or "installed" as a pre-commit hook with:

```
pre-commit install
```

where it will be run after every commit:

```
(py38) alex@enterprise:~/python/pyvista$ git commit -am "implement mypy and update yml"
black....................................................................Passed
isort....................................................................Passed
flake8...................................................................Passed
codespell................................................................Passed
pydocstyle...........................................(no files to check)Skipped
mypy.................................................(no files to check)Skipped
```

It only checks the files staged for the commit and is therefore wicked fast, something we want if we want contributors to want to use this tool. Best part is if something needs to be autoformatted by `black` or `isort`, you run the tool, it modifies the files and tells you it failed, and then you simply commit again.

It's super automatic and means we won't have to keep coming back to failed PRs because we had a character of whitespace on L2087.

### Implementations

This PR removes:
- The `requirements_style.txt` and `requirements_static.txt` and these versions are now tracked in `.pre-commit-config.yaml`.
- Several sections have been removed from the `Makefile`. "There should be one-- and preferably only one --obvious way to do it."

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -264,40 +254,48 @@ leverage multiple processes. Example usage:
...
+This way, it's not possible for you to push code that fails the style
+checks. For example, each commit automatically checks that you meet the style
+requirements::
 
-   cd doc
-   make clean
-   make doctest
-   make html -b linkcheck
+  $ pre-commit install
+  $ git commit -am "added my cool feature"
```
Comment (tkoyama010) [inline CONTRIBUTING.rst:293]:
```suggestion
  $ git commit -m "added my cool feature"
```

**Review by tkoyama010 — APPROVED** [review]
Thanks for this. LGTM with a small fix.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -264,40 +254,48 @@ leverage multiple processes. Example usage:
...
-   cd doc
-   make clean
-   make doctest
-   make html -b linkcheck
+  $ pre-commit install
+  $ git commit -m "added my cool feature"
+  black....................................................................Passed
+  isort....................................................................Passed
+  flake8...................................................................Passed
+  codespell................................................................Passed
```
Comment (adeak) [inline CONTRIBUTING.rst:298]:
```suggestion

The actual installation of the environment happens before the first commit 
following ``pre-commit install``. This will take a bit longer, but subsequent
commits will only trigger the actual style checks.
```

**Review by adeak — COMMENTED** [review]
Thanks for this, I've been meaning to do that for a while but never got around to it.

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -264,40 +254,48 @@ leverage multiple processes. Example usage:
...
-   export PYVISTA_OFF_SCREEN=True
    python -m pytest -v --doctest-modules pyvista
 
-Run documentation testing by running:
+Build the documentation with:
 
 .. code:: bash
 
-   make
+   make -C doc html
```
Comment (dcbr) [inline CONTRIBUTING.rst:269]: On Windows I have to use `python -msphinx -M html . _build` because Makefiles are not supported (as you mentioned here earlier). Maybe good to add such a note here as well now that everything else is made cross-platform?

**Review by dcbr — COMMENTED** [review]
Really great addition which will save a lot of time :)

I haven't tested it locally, but what happens if multiple formatting steps would fail (e.g. both `black` and `isort` have to modify files)? Do you have to commit again for each separate step or are all changes performed at once, requiring only a single new commit attempt afterwards?
Also does the error message explicitly say to try and commit again, because otherwise it might be quite confusing as to why commiting failed exactly?

File: pyvista/ext/plot_directive.py [hunk 4]
```diff
@@ -99,6 +99,7 @@
 import shutil
 import textwrap
 import traceback
+from typing import Any, Dict
```
Comment (tkoyama010) [inline pyvista/ext/plot_directive.py:102]:
```suggestion
```

**Comment by MatthewFlamm** [comment]
> if multiple formatting steps would fail (e.g. both `black` and `isort` have to modify files)? Do you have to commit again for each separate step or are all changes performed at once, requiring only a single new commit attempt afterwards?

The attempted commit is aborted and changes are made to the local workspace, in the case of black and isort.  You have to `git add` the changes and then commit again.

One additional benefit of pre-commit is that it runs against the code being committed, which prevents errors when you forget to stage changes.

File: CONTRIBUTING.rst (hunk 3 above)
Comment (akaszynski) [reply CONTRIBUTING.rst:269]: Added in b5fa2f74c97f4ed434919f5265f739ed0ad2e400

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  - id: isort
+
+- repo: https://gitlab.com/PyCQA/flake8
+  rev: 3.9.2
+  hooks:
+  - id: flake8
+    additional_dependencies: [
+      "flake8-black==0.3.2",
+      "flake8-isort==4.1.1",
+      "flake8-quotes==3.3.1",
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply .pre-commit-config.yaml:19]: added in a69e1eba08b6e5f73cc9d7863b168f47f2efd8d7

**Comment by akaszynski** [comment]
Added in a few other pre-commit checks:
- Disable committing to `main`
- Flag introduction of `breakpoints` (I've done that)
- Check merge conflicts
- Validate github yml schema (this will save us from failed runs for boneheaded mistakes)
